command/agent/config/config.go

```diff
@@ -169,14 +171,249 @@ func NewConfig() *Config {
+			if c.AutoAuth == nil {
+				return fmt.Errorf("cache.use_auto_auth_token is true but auto_auth not configured")
+			}
+			if c.AutoAuth != nil && c.AutoAuth.Method != nil && c.AutoAuth.Method.WrapTTL > 0 {
+				return fmt.Errorf("cache.use_auto_auth_token is true and auto_auth uses wrapping")
+			}
+		}
+	}
+
+	if c.APIProxy != nil {
+		if len(c.Listeners) < 1 {
+			return fmt.Errorf("configuring the api_proxy requires at least 1 listener to be defined")
+		}
+
+		if c.APIProxy.UseAutoAuthToken {
+			if c.AutoAuth == nil {
+				return fmt.Errorf("api_proxy.use_auto_auth_token is true but auto_auth not configured")
+			}
+			if c.AutoAuth != nil && c.AutoAuth.Method != nil && c.AutoAuth.Method.WrapTTL > 0 {
+				return fmt.Errorf("api_proxy.use_auto_auth_token is true and auto_auth uses wrapping")
+			}
+		}
+	}
+
+	if c.AutoAuth != nil {
+		if len(c.AutoAuth.Sinks) == 0 &&
+			(c.APIProxy == nil || !c.APIProxy.UseAutoAuthToken) &&
+			len(c.Templates) == 0 {
+			return fmt.Errorf("auto_auth requires at least one sink or at least one template or api_proxy.use_auto_auth_token=true")
+		}
+	}
+
+	return nil
+}
+
 // LoadConfig loads the configuration at the given path, regardless if
-// its a file or directory.
+// it's a file or directory.
 func LoadConfig(path string) (*Config, error) {
 	fi, err := os.Stat(path)
 	if err != nil {
 		return nil, err
 	}
 
+	if fi.IsDir() {
+		return LoadConfigDir(path)
+	}
+	return LoadConfigFile(path)
+}
+
+// LoadConfigDir loads the configuration at the given path if it's a directory
+func LoadConfigDir(dir string) (*Config, error) {
+	f, err := os.Open(dir)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, err
+	}
+	if !fi.IsDir() {
+		return nil, fmt.Errorf("configuration path must be a directory: %q", dir)
+	}
+
+	var files []string
+	err = nil
+	for err != io.EOF {
+		var fis []os.FileInfo
+		fis, err = f.Readdir(128)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+
+		for _, fi := range fis {
+			// Ignore directories
+			if fi.IsDir() {
+				continue
+			}
+
+			// Only care about files that are valid to load.
+			name := fi.Name()
+			skip := true
+			if strings.HasSuffix(name, ".hcl") {
+				skip = false
+			} else if strings.HasSuffix(name, ".json") {
+				skip = false
+			}
+			if skip || isTemporaryFile(name) {
+				continue
+			}
+
+			path := filepath.Join(dir, name)
+			files = append(files, path)
+		}
+	}
+
+	result := NewConfig()
+	for _, f := range files {
+		config, err := LoadConfigFile(f)
+		if err != nil {
+			return nil, fmt.Errorf("error loading %q: %w", f, err)
+		}
+
+		if result == nil {
+			result = config
+		} else {
+			result = result.Merge(config)
+		}
+	}
+
+	return result, nil
+}
+
+// isTemporaryFile returns true or false depending on whether the
+// provided file name is a temporary file for the following editors:
+// emacs or vim.
+func isTemporaryFile(name string) bool {
```

Comment:
This is a little weird to me, but we use it for Vault's config parsing so I thought I'd keep consistency.